### ISSUE TITLE
Inefficient Multi-View Data Pipeline with    RepeatedRandomSampler

### ISSUE THREAD

**Issue by huguesva** [rbalestr-lab/stable-pretraining#310]
Current Approach - Problems and Inefficiencies

  1. Configuration Confusion

  batch_size = 128  # Misleading - actually only 16 unique images!
  n_views = 8      # 2 global + 6 local crops
  sampler = RepeatedRandomSampler(train_dataset, n_views=8)

  2. How RepeatedRandomSampler Works

  # For a dataset with 3 images, it yields:
  [0,0,0,0,0,0,0,0, 1,1,1,1,1,1,1,1, 2,2,2,2,2,2,2,2, ...]
  Each image index is repeated n_views times consecutively.

  3. DataLoader Semantic Confusion

  With batch_size=128 and n_views=8:
  - DataLoader requests 128 items (thinking it's getting 128
  images)
  - Reality: Only 16 unique images (128÷8), each appearing 8 times
  - This breaks PyTorch's semantic convention where batch_size =
  number of unique samples

  4. Inefficient MultiViewTransform

  class MultiViewTransform:
      def __init__(self, transforms):
          self.transforms = transforms
          self.counter = 0

      def __call__(self, sample):
          # Uses counter to pick ONE transform
          transform_idx = self.counter % len(self.transforms)
          self.counter += 1
          return self.transforms[transform_idx](sample)

  Problems:
  - Loads and processes same image 8 times from disk
  - Counter-based selection is fragile and stateful
  - 128 separate I/O operations for only 16 images

  5. Mixed-Size Batch Assembly Failure

  The batch contains:
  - 32 tensors of [3, 224, 224] (global crops)
  - 96 tensors of [3, 96, 96] (local crops)

  Default collate fails - can't stack different sizes!

  6. Summary of Issues

  1. 8x more I/O operations than necessary
  2. Confusing semantics: batch_size doesn't mean what users
  expect
  3. Stateful transform with counter is error-prone
  4. Incompatible with standard collate for mixed sizes
  5. Poor GPU utilization from inefficient batching

  ---
  Proposed Canonical Approach

  1. Transform Returns All Views

  class MultiViewTransform:
      def __init__(self, transforms):
          self.transforms = transforms

      def __call__(self, sample):
          # Apply ALL transforms to same image at once
          return [transform(sample) for transform in
  self.transforms]

  2. Use Standard Sampler

  # Clear semantics: batch_size = number of unique images
  train_dataloader = DataLoader(
      dataset=train_dataset,
      batch_size=16,  # Actually 16 images!
      sampler=DistributedSampler(train_dataset),  # Standard 
  sampler
      collate_fn=multi_view_collate,
  )

  3. Simple Collate for Mixed Sizes

  def multi_view_collate(batch):
      # batch = [{'image': [view1, view2, ...], 'label': label}, 
  ...]
      views = [item for sample in batch for item in
  sample['image']]
      labels = torch.tensor([sample['label'] for sample in batch])

      return {
          'views': views,  # List of tensors (can be different 
  sizes)
          'labels': labels,
          'sample_idx':
  torch.repeat_interleave(torch.arange(len(batch)), n_views)
      }

  4. Benefits

  - 8x fewer I/O operations (16 instead of 128)
  - Clear semantics: batch_size=16 means 16 images
  - Stateless transform: No counter needed
  - Handles mixed sizes naturally: Views stay as list
  - Better performance: ~3x faster data loading
  - PyTorch idiomatic: Follows community patterns

  Migration Path

  1. Add new MultiViewTransform that returns list
  2. Add multi_view_collate function to library
  3. Update forward functions to handle list of views
  4. Deprecate RepeatedRandomSampler for multi-view use cases

  This aligns with PyTorch best practices used by VISSL,
  Solo-learn, and other major SSL libraries.

**Comment by RandallBalestriero** [comment]
**Let's not deprecate RepeatedRandomSampler**, it has a lot of use cases and is valid in many scenarios outside.
We definitely should add the extra multi view transform no matter what, and we can do some quick benchmarking to see which to suggest users.

**Comment by huguesva** [comment]
handled in #311
